mlflow/store/local_artifact_repo.py

```diff
@@ -72,7 +72,7 @@ def list_artifacts(self, path=None):
             path = os.path.normpath(path)
         list_dir = os.path.join(self.artifact_dir, path) if path else self.artifact_dir
         if os.path.isdir(list_dir):
-            artifact_files = list_all(list_dir, full_path=True)
+            artifact_files = list_all(list_dir, full_path=True) + [list_dir]
```

Comment:
The spark UDF test (`pytest tests/pyfunc/test_spark.py::test_spark_udf --large`) was failing due to leaked mlruns directory from another test run (we should fix that separately), see [this travis log](https://travis-ci.org/mlflow/mlflow/jobs/562924587). The failure revealed that we currently have a bug/edge case in how we handle downloading artifacts from empty directories (in ArtifactRepository.download_artifacts, we assume that an empty result from list_artifacts implies that an artifact path is a file, but it could actually be an empty directory, e.g. the `mlruns` directory in the above failure). I changed this code to get the test to pass on the PR, but this is probably not the right solution (will think about this more tomorrow)

Reply:
I don't understand why do you want to add the list_dir to the result. Can you elaborate on what caused the error and how does this fixes it? From looking at it briefly I don't think the list_dir should be returned. If anything, I don't understand why list artifacts pointing to a file does not return the file itself.

Reply:
Yeah agreed, adding `list_dir` is just a hack to get the tests to pass against the file-based artifact repo, because it allows us to distinguish between listing an empty directory & a file.

Agree that it'd make sense for list_artifacts pointing a file to return the file itself, but that's a broader change that requires modifications across the different artifact repository implementations, so I just worked around this for now by removing the problematic empty `mlruns` directory in the `projects` tests that create it.